File: Unity/Assets/Scripts/Systems/Shelter.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace Vacancy
{
    public sealed class ShelterState
    {
        public bool Unlocked;
        public bool DefenseActive;
        public readonly Dictionary<string, int> Stock = new Dictionary<string, int>();
        public float Integrity = 100f;
        public bool Powered = true;
        public int DaysHeld;
        public string LastShortage;
    }

    public static class Shelter
    {
        public static ShelterState Create()
        {
            var state = new ShelterState();
            foreach (var pair in GameConfig.ShelterItems)
            {
                state.Stock[pair.Key] = pair.Value.StartingStock;
            }

            return state;
        }

        public static List<string> ItemIds()
        {
            var ids = new List<string>();
            foreach (var pair in GameConfig.ShelterItems) ids.Add(pair.Key);
            return ids;
        }

        public static int GetStock(GameState state, string itemId)
        {
            if (state.Shelter == null) return 0;
            return state.Shelter.Stock.TryGetValue(itemId, out var qty) ? qty : 0;
        }

        public static void UnlockShelterSystems(GameState state)
        {
            if (state.Shelter == null || state.Shelter.Unlocked) return;
            state.Shelter.Unlocked = true;
            state.AddLog(
                "You inventory what the hotel actually has: water, food, fuel, medicine, and lumber.");
        }

        public static void ActivateDefense(GameState state)
        {
            if (state.Shelter == null || state.Shelter.DefenseActive) return;
            state.Shelter.DefenseActive = true;
            state.AddLog("Barricades go up over the ground-floor windows.");
        }

        public static int CountOccupants(GameState state)
        {
            int inRooms = 0;
            foreach (var room in state.Rooms)
            {
                if (room.Status == "occupied") inRooms++;
            }

            int staff = (state.BobHired ? 1 : 0) + (state.MaryHired ? 1 : 0);
            return inRooms + staff + 1;
        }

        static bool Spend(GameState state, string itemId, int qty)
        {
            int have = GetStock(state, itemId);
            int used = Math.Min(have, qty);
            state.Shelter.Stock[itemId] = have - used;
            return used == qty;
        }

        public static void ProcessDailyShelter(GameState state)
        {
            var shelter = state.Shelter;
            if (shelter == null || !shelter.Unlocked) return;

            int occupants = CountOccupants(state);
            var shortages = new List<string>();

            if (!Spend(state, "water", occupants * (int)GameConfig.WaterPerPersonPerDay))
            {
                shortages.Add("water");
            }

            if (!Spend(state, "food", occupants * (int)GameConfig.FoodPerPersonPerDay))
            {
                shortages.Add("food");
            }

            if (!Spend(state, "fuel", (int)GameConfig.FuelPerDay))
            {
                shortages.Add("fuel");
                shelter.Powered = false;
            }
            else
            {
                shelter.Powered = true;
            }

            if (shortages.Count > 0)
            {
                shelter.LastShortage = string.Join(", ", shortages);
                state.Reputation = Math.Max(0, state.Reputation - shortages.Count * 2);
                state.AddLog($"Ran short on {shelter.LastShortage} for {occupants} people. Morale drops.");
            }
            else
            {
                shelter.LastShortage = null;
                shelter.DaysHeld += 1;
            }

            if (shelter.DefenseActive)
            {
                shelter.Integrity = Math.Max(0, shelter.Integrity - GameConfig.NightlyIntegrityLoss);
                if (shelter.Integrity <= 0)
                {
                    state.Reputation = Math.Max(0, state.Reputation - 6);
                    state.AddLog("The barricades failed overnight. Something got in.");
                    shelter.Integrity = 10;
                }
                else if (shelter.Integrity < GameConfig.DangerIntegrity)
                {
                    state.AddLog(
                        $"Barricades down to {Math.Round(shelter.Integrity)}%. Reinforce them with lumber.");
                }
            }
        }

        public static bool ReinforceBarricades(GameState state, int lumber = 1)
        {
            var shelter = state.Shelter;
            if (shelter == null || !shelter.Unlocked) return false;
            if (shelter.Integrity >= 100)
            {
                state.AddLog("The barricades are already solid.");
                return false;
            }

            int available = Math.Min(lumber, GetStock(state, "lumber"));
            if (available <= 0)
            {
                state.AddLog("No lumber left. Order more before nightfall.");
                return false;
            }

            state.Shelter.Stock["lumber"] -= available;
            float gain = available * GameConfig.IntegrityPerLumber;
            shelter.Integrity = Math.Min(100, shelter.Integrity + gain);
            state.AddLog(
                $"Reinforced the barricades with {available} lumber ({Math.Round(shelter.Integrity)}%).");
            return true;
        }

        public static string HudSummary(GameState state)
        {
            var shelter = state.Shelter;
            if (shelter == null || !shelter.Unlocked) return "";

            var sb = new StringBuilder();
            bool first = true;
            foreach (var pair in GameConfig.ShelterItems)
            {
                if (!first) sb.Append(" · ");
                first = false;
                sb.Append(pair.Value.Label).Append(": ").Append(GetStock(state, pair.Key));
            }

            sb.Append(" · Barricades: ").Append(Math.Round(shelter.Integrity)).Append("%");
            if (!shelter.Powered) sb.Append(" · GENERATOR DOWN");
            return sb.ToString();
        }
    }
}

File: js/shelter.js
import { CONFIG } from "./config.js";
import { addLog } from "./state.js";

/**
 * Shelter management — the late-game evolution of the supply closet.
 *
 * Dormant until the story unlocks it, so early play is untouched. After that,
 * headcount burns water and food, the generator burns fuel, and the barricades
 * lose integrity every night you do not maintain them.
 */

export function createShelterState() {
  const stock = {};
  for (const id of Object.keys(CONFIG.shelterItems)) {
    stock[id] = CONFIG.shelterItems[id].startingStock;
  }
  return {
    /** Ordering and tracking becomes available when the city goes dark. */
    unlocked: false,
    /** Nightly attrition and threat checks start at collapse. */
    defenseActive: false,
    stock,
    integrity: 100,
    powered: true,
    /** Days survived with everyone fed and the doors holding. */
    daysHeld: 0,
    lastShortage: null,
  };
}

export function shelterItemIds() {
  return Object.keys(CONFIG.shelterItems);
}

export function getShelterStock(state, itemId) {
  return state.shelter?.stock?.[itemId] ?? 0;
}

export function unlockShelterSystems(state) {
  if (!state.shelter || state.shelter.unlocked) return;
  state.shelter.unlocked = true;
  addLog(
    state,
    "You inventory what the hotel actually has: water, food, fuel, medicine, lumber."
  );
}

export function activateDefense(state) {
  if (!state.shelter || state.shelter.defenseActive) return;
  state.shelter.defenseActive = true;
  addLog(state, "Barricades go up over the ground-floor windows.");
}

/** Everyone the shelter is currently keeping alive. */
export function countOccupants(state) {
  const inRooms = state.rooms.filter((r) => r.status === "occupied").length;
  const staff = (state.bobHired ? 1 : 0) + (state.maryHired ? 1 : 0);
  return inRooms + staff + 1;
}

function spend(state, itemId, qty) {
  const have = getShelterStock(state, itemId);
  const used = Math.min(have, qty);
  state.shelter.stock[itemId] = have - used;
  return used === qty;
}

/** Called once per day rollover. */
export function processDailyShelter(state) {
  const shelter = state.shelter;
  if (!shelter?.unlocked) return;

  const occupants = countOccupants(state);
  const use = CONFIG.shelterUse;
  const shortages = [];

  if (!spend(state, "water", occupants * use.waterPerPersonPerDay)) {
    shortages.push("water");
  }
  if (!spend(state, "food", occupants * use.foodPerPersonPerDay)) {
    shortages.push("food");
  }
  if (!spend(state, "fuel", use.fuelPerDay)) {
    shortages.push("fuel");
    shelter.powered = false;
  } else {
    shelter.powered = true;
  }

  if (shortages.length) {
    shelter.lastShortage = shortages.join(", ");
    state.reputation = Math.max(0, state.reputation - shortages.length * 2);
    addLog(
      state,
      `Ran short on ${shelter.lastShortage} for ${occupants} people. Morale drops.`
    );
  } else {
    shelter.lastShortage = null;
    shelter.daysHeld += 1;
  }

  if (shelter.defenseActive) {
    const loss = CONFIG.shelterDefense.nightlyIntegrityLoss;
    shelter.integrity = Math.max(0, shelter.integrity - loss);
    if (shelter.integrity <= 0) {
      state.reputation = Math.max(0, state.reputation - 6);
      addLog(state, "The barricades failed overnight. Something got in.");
      shelter.integrity = 10;
    } else if (shelter.integrity < CONFIG.shelterDefense.dangerIntegrity) {
      addLog(
        state,
        `Barricades down to ${Math.round(shelter.integrity)}%. Reinforce them with lumber.`
      );
    }
  }
}

/** Spend lumber to bring the barricades back up. Returns true if anything changed. */
export function reinforceBarricades(state, lumber = 1) {
  const shelter = state.shelter;
  if (!shelter?.unlocked) return false;
  if (shelter.integrity >= 100) {
    addLog(state, "The barricades are already solid.");
    return false;
  }
  const available = Math.min(lumber, getShelterStock(state, "lumber"));
  if (available <= 0) {
    addLog(state, "No lumber left. Order more before nightfall.");
    return false;
  }

  state.shelter.stock.lumber -= available;
  const gain = available * CONFIG.shelterDefense.integrityPerLumber;
  shelter.integrity = Math.min(100, shelter.integrity + gain);
  addLog(
    state,
    `Reinforced the barricades with ${available} lumber (${Math.round(shelter.integrity)}%).`
  );
  return true;
}

export function shelterHudSummary(state) {
  const shelter = state.shelter;
  if (!shelter?.unlocked) return "";
  const parts = shelterItemIds().map(
    (id) => `${CONFIG.shelterItems[id].label}: ${getShelterStock(state, id)}`
  );
  parts.push(`Barricades: ${Math.round(shelter.integrity)}%`);
  if (!shelter.powered) parts.push("GENERATOR DOWN");
  return parts.join(" · ");
}
